File: 프로그래머스/lv3/42579. 베스트앨범/베스트앨범.py
from collections import defaultdict


def solution(genres, plays):
    answer = []
    total_play = defaultdict(int)
    genre_list = defaultdict(list)
    
    for i, (play, genre) in enumerate(zip(plays, genres)):
        total_play[genre] += play
        genre_list[genre].append((play, i))
    
    for k in sorted(total_play, key=lambda x: total_play[x], reverse=True):
        for g in sorted(genre_list[k], key=lambda x: (-x[0], x[1]))[:2]:
            answer.append(g[1])
        
    return answer

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.java
import java.util.*;
import java.util.stream.*;

class Solution {

    public int[] solution(String[] genres, int[] plays) {
        Map<String, Integer> totalPlay = new HashMap<>();
        Map<String, List<int[]>> genreList = new HashMap<>();

        for (int i = 0; i < genres.length; i++) {
            totalPlay.merge(genres[i], plays[i], Integer::sum);
            genreList.computeIfAbsent(genres[i], k -> new ArrayList<>())
                     .add(new int[]{plays[i], i});
        }

        return totalPlay.entrySet().stream()
                .sorted(Map.Entry.<String, Integer>comparingByValue().reversed())
                .flatMap(e -> genreList.get(e.getKey()).stream()
                        .sorted(Comparator.comparingInt((int[] a) -> -a[0])
                                         .thenComparingInt(a -> a[1]))
                        .limit(2)
                        .map(a -> a[1]))
                .mapToInt(Integer::intValue)
                .toArray();
    }
}